lib/clearance/user.rb

```diff
@@ -75,13 +75,13 @@ def reset_remember_token!
     def update_password(new_password)
       self.password_changing = true
       self.password = new_password
+      valid?
 
-      if valid?
+      if errors[:password].blank?
         self.confirmation_token = nil
         generate_remember_token
+        save(validate: false)
```

Comment:
Can't I call `user.update_attributes(password: new_password, confirmation_token: user.confirmation_token, remember_token: user.remember_token)`. It's odd looking but I think it would work, right?

> This would be a breaking change, but since it seems like the save/validate lifecycle of users/passwords is different, one thing you could do is split out a new table/model to represent authentication tokens and make them belong to a user. That way, you could expire/delete old tokens and create new ones without touching the user/model table and its maze of validations.

This is already in the plans for addressing #465. The plan for that is:

1. Extract password reset to a service and make the methods in `Clearance::User` passthroughs to that service, deprecate direct calls.
2. Expose a new service that splits out password resets, make this the default for newly installed apps in the initializer.
3. Deprecate the old service.